components/calendar/src/date.rs

```diff
@@ -239,36 +240,66 @@ impl<A: AsCalendar> Date<A> {
     /// Add a `duration` to this date, mutating it
     #[doc(hidden)] // unstable
     #[inline]
-    pub fn add(&mut self, duration: DateDuration) {
-        self.calendar
+    pub fn try_add_with_options(
+        &mut self,
+        duration: types::DateDuration,
+        options: DateAddOptions,
+    ) -> Result<(), DateError> {
+        let inner = self
+            .calendar
             .as_calendar()
-            .offset_date(&mut self.inner, duration)
+            .add(&self.inner, duration, options)?;
+        self.inner = inner;
+        Ok(())
     }
 
     /// Add a `duration` to this date, returning the new one
     #[doc(hidden)] // unstable
     #[inline]
-    pub fn added(mut self, duration: DateDuration) -> Self {
-        self.add(duration);
-        self
+    pub fn try_added_with_options(
```

Comment:
suggestion: I don't like that `is_negative` is a field in `DateDuration`. I'd rather keep durations positive, and add a subtract method on `Date` (`Calendar::add` can take `is_negative` as an additional argument).

Reply:
Then we need something to be able to return negative durations in `until()`.

I think `is_negative` is a good model for durations.

Reply:
hmm. maybe `until` should return `Result<Duration, Duration>`. it is somewhat likely that a caller expects the result to be positive (otherwise they might have called `since`), and they might even forget to check the `is_negative` flag of the result.

Reply:
I think that's an awkward API, not to mention that Until already needs to return an error. I don't really see the problem with negative durations, they're a thing you need to handle in many cases.

Reply:
I don't have problem with negative durations, I do have a problem with the `is_negative` flag. in the return position it has to be explicitly checked which is easy to forget

Reply:
I don't understand: if you wish to correctly interpret a duration _all_ of the duration fields need to be checked. Why is `is_negative` special?